pure/pure.1
.TH Pure 1 "September 2014" "Pure" "Pure Manual"
.SH NAME
pure \- the Pure interpreter
.SH SYNOPSIS
\fBpure\fP [\fIoptions\fP ...] [-x] \fIscript\fP [\fIargs\fP ...]
.br
\fBpure\fP [\fIoptions\fP ...] [-b|-c|-i] [\fIscript\fP ...] [-- \fIargs\fP ...]
.SH OPTIONS
.TP
.B -b
Batch mode (execute the given scripts and exit).
.TP
.B -c
Batch compilation (compile the scripts to a native binary).
.TP
.B --check
Syntax check only, do not execute code or produce any output files.
.TP
\fB--ctags\fP, \fB--etags\fP
Create a tags file in ctags (vi) or etags (emacs) format.
.TP
.BI --disable= optname
Disable source option (conditional compilation).
.TP
.B --eager-jit
Enable eager JIT compilation. This requires LLVM 2.7 or later, otherwise this
flag will be ignored.
.TP
.BI --enable= optname
Enable source option (conditional compilation).
.TP
.BI --escape= char
Interactive commands are prefixed with the specified character. Permitted
prefixes are: !$%&*,:<>@\\|.
.TP
\fB-fPIC\fP, \fB-fpic\fP
Create position-independent code (batch compilation).
.TP
.B -g
Enable symbolic debugging.
.TP
\fB--help\fP, \fB-h\fP
Print help message and exit.
.TP
.B -i
Interactive mode (read commands from stdin after sourcing the given scripts,
if any).
.TP
.BI -I directory
Add a directory to be searched for included source scripts.
.TP
.BI -L directory
Add a directory to be searched for dynamic libraries.
.TP
.BI -l libname
Library to be linked in batch compilation.
.TP
.BI -m opt = val
Add llc machine options in batch compilation.
.TP
.BI --main= name
Name of main entry point in batch compilation.
.TP
.B --noediting
Disable command-line editing.
.TP
\fB--noprelude\fP, \fB-n\fP
Do not load the prelude.
.TP
.B --norc
Do not run the interactive startup files.
.TP
.BI -o filename
Output filename for batch compilation.
.TP
.B -q
Quiet startup (suppresses sign-on message in interactive mode).
.TP
.BI -T filename
Tags file to be written by --ctags or --etags.
.TP
.B --texmacs
Run Pure inside TeXmacs.
.TP
.B -u
Do not strip unused functions in batch compilation.
.TP
.BR -v [\fIlevel\fP]
Set verbosity level.
.TP
.B --version
Print version information and exit.
.TP
.B -w
Enable compiler warnings about various dubious constructs and backward
compatibility issues.
.TP
.B -x
Execute a script with the given command line arguments.
.TP
.B --
Stop option processing and pass the remaining command line arguments in the
.B argv
variable.
.SH DESCRIPTION
Pure is a modern-style functional programming language based on term
rewriting. Pure programs are basically collections of equational rules used to
evaluate expressions in a symbolic fashion by reducing them to normal form.
.PP
.B pure
is the Pure interpreter. The interpreter has an LLVM backend which
JIT-compiles Pure programs to machine code, hence programs run blazingly fast
and interfacing to C modules is easy, while the interpreter still provides a
convenient, fully interactive environment for running Pure scripts and
evaluating expressions.
.PP
Basically, to run a Pure script with the interpreter, just specifiy the
filename of the script on the command line (optionally followed by additional
arguments to be passed to the script). You can also invoke the interpreter
without arguments, in which case it starts up in interactive mode. Or you can
compile one or more scripts to a native executable (as well as a number of
other recognized LLVM and native code formats) with the
.B -c
option (use
.B -o
to specify the output filename); this requires that you have some parts of the
LLVM toolchain (llc, opt) as well as gcc installed. Last but not least, you
can create a tags file for use with
.BR emacs (1)
or
.BR vi (1)
with the
.B --etags
and
.B --ctags
options.
.PP
The full Pure manual can be read inside the interpreter, by typing
.B help
on the interpreter's command line. This requires
.BR w3m (1)
to work. (You can also set the name of another html browser to use with the
.B PURE_HELP
or the
.B BROWSER
environment variable, or read the manual online at the Pure website.)
.SH FILES
.TP
.B a.out
Default output filename for batch compilation (-c).
.TP
.B ~/.pure_history
Interactive command history.
.TP
\fB~/.purerc\fP, \fB.purerc\fP, \fB.pure\fP
Interactive startup files. The latter is usually a dump from a previous
interactive session.
.TP
.B prelude.pure
Standard prelude. If available, this script is loaded before any other
definitions, unless
.B -n
was specified.
.SH ENVIRONMENT
.TP
\fBCC\fP, \fBCXX\fP
C and C++ compiler used by the Pure batch compiler (\fBpure -c\fP) to compile
and link native executables. Defaults to \fBgcc\fP and \fBg++\fP, respectively.
.TP
.B BROWSER
If the
.B PURE_HELP
variable is not set (see below), this specifies a colon-separated list of
browsers to try for reading the online documentation. More information about
the \fBBROWSER\fP variable can be found at
\fIhttp://www.catb.org/~esr/BROWSER/\fP.
.TP
.B PURELIB
Directory to search for library scripts, including the prelude. If
.B PURELIB
is not set, it defaults to some location specified at installation time.
.TP
.B PURE_EAGER_JIT
Enable eager JIT compilation (same as \fB--eager-jit\fP).
.TP
.B PURE_ESCAPE
If set, interactive commands are prefixed with the first character in the
value of this variable (same as \fB--escape\fP). If the variable is empty, or
its first character is not a permitted prefix, Pure reverts to the default of
no prefix.
.TP
.B PURE_HELP
Browser to use to read online documentation in the interpreter. By default,
.BR w3m (1)
is used.
.TP
.B PURE_INCLUDE
Additional directories (in colon-separated format) to be searched for included
scripts.
.TP
.B PURE_LIBRARY
Additional directories (in colon-separated format) to be searched for dynamic
libraries.
.TP
.B PURE_MORE
Shell command to be used for paging through output of the
.B show
command, when the interpreter runs in interactive mode.
.B PURE_LESS
does the same for evaluation results printed by the interpreter.
.TP
.B PURE_PS
Command prompt used in the interactive command loop (">\ " by default).
.TP
.B PURE_STACK
Maximum stack size in kilobytes (0 = unlimited). A reasonable default is
provided (currently this is always 8192K - 128K for interpreter and runtime,
which should work on most modern PCs). If you're still getting segfaults due
to stack overflow then you'll either have to reduce this value or increase the
actual stack space available to programs.
.SH LICENSE
LGPL V3 or later. See the accompanying COPYING file for details.
.SH AUTHOR
Albert Graef <aggraef@gmail.com>, Dept. of Computer Music, Johannes Gutenberg
University of Mainz, Germany.
.SH SEE ALSO
You can find the latest releases, as well as the complete manual and the
mailing list at the Pure website, \fIhttps://agraef.github.io/pure-lang\fP.


pure/NEWS
** Pure 0.67 2018-03-18

Yet another maintenance release, with a few updates and fixes to make the
source compile with the latest gcc versions, a new Windows/msys2 port, a new
syntax highlighting file (Notepad++), and added flycheck support for Emacs. In
order to support the latter, the interpreter has a new --check option which
checks the syntax of the given scripts in batch mode without executing them or
producing any output files. Other user-visible changes:

- The direct Faust interface has been fixed up to work with the latest Faust
  versions (2.5.0+) again. (Issue #12)

- The -fPIC flag is now added automatically in batch compilation on platforms
  which need it, making batch compiles a little more convenient. (Issue #14)

Finally, Pure has moved to Github (https://agraef.github.io/pure-lang), so
please update your bookmarks. (The old Bitbucket repository still exists for
the time being and older downloads can still be found there, but it is a
read-only archive now and will not be updated any longer.)

** Pure 0.66 2017-03-04

Another maintenance release which sports a few additional bugfixes in Emacs
Pure mode. Also, --version now reports the git revision if it is available at
compile time, and a new example was added (examples/lucid.pure).

The biggest change in this release is in the way the compiler treats the
declaration of binary minus in custom namespaces. The compiler now handles
this by manufacturing a corresponding unary minus operator and an accompanying
"neg" function symbol in the same namespace, so that the overloaded "-"
operator works in exactly the same manner as the default one. Note that no
special action is needed to define the unary variant of the operator; you only
declare the binary operator (like it's done in the prelude for the standard
'-' operator), and the compiler does all the necessary incantations to make
this work.

This finally makes it possible to completely override all the arithmetic
operators *including* unary minus in a custom namespace (this is also used in
the lucid.pure example, so you can see it in action there). Previously, if
my::- was declared as a binary minus operator in namespace my, using it as a
unary minus operator would always provoke an error message from the parser.

** Pure 0.65 2017-02-23

Pure 0.64 turned out to be an incredibly stable release, but a maintenance
release is finally in order, to pick up the (very few) changes which have been
piling up in the source repository since 0.64:

* Fixed a crash bug in the compiler. You can tell that a compiler is getting
  fairly mature if compiler bugs become as rare and obscure as this one. ;-)

* Minor bugfixes in the debugger, expression printer and the embedded runtime.

* Minor bugfixes in ATS compilation, also added some more ATS examples
  contributed by Barry Schwartz.

* Pure mode supports the latest Emacs versions now (24+25).

* Documentation updates.

As usual, details can be found in the ChangeLog.

Another (unfortunate) newsworthy item is that dragonegg, gcc's LLVM plugin,
has obviously bitten the dust, which prompted some minor changes in the
bitcode examples; by default we use clang there for C/C++ bitcode compilation
now. Alas, dragonegg's untimely death also means that there's no easy pathway
for Fortran LLVM bitcode compilation any more, and thus Pure's Fortran
inlining facility has become rather useless now. If you really need this, you
will have to stick to gcc 4.6, which seems to be the last gcc version still
supported by dragonegg. Otherwise we recommend just using the dynamic library
loading mechanism instead, if you need to interface to Fortran code.

Finally, note that we switched to git quite a while ago already, so the
repository is now at https://agraef@bitbucket.org/purelang/pure-lang.git
(read-only) and git@bitbucket.org:purelang/pure-lang.git (developers only).

** Pure 0.64 2014-10-28

This release features improvements in the batch compiler (issue #30), as well
as support for Perl-style regular expressions (issue #20).

* As proposed by Alastair Pharo, the batch compiler now uses some linker magic
  in order to locate imported libraries outside of Pure's library path, so
  that linking against system libraries in batch-compiled modules should be
  much easier now.

* At long last, Pure gets support for Perl-style regular expressions via
  libpcre (http://www.pcre.org/). This is still experimental and currently
  disabled by default. Configuring with --with-pcre enables it. Note that at
  present this is only available as a compile-time switch; to change the regex
  library you need to recompile the interpreter.

** Pure 0.63 2014-10-22

Barry Schwartz contributed support for inlining ATS (an interesting ML-like
functional language with program verification capabilities, see
http://www.ats-lang.org/). Thanks, Barry!

Also, this release sports some fixes in the built-in Faust interface (prompted
by recent changes in the Faust compiler's treatment of empty group and control
labels), as well as some portability fixes (mostly for Mac OS X and Windows).

The ChangeLog is a bit short on details this time, so please refer to the
commit history of the Mercurial repository if you need to find out about all
the gory details.

** Pure 0.62 2014-09-16

New -mopt=val option of the batch compiler which allows the corresponding llc
machine options (-march, -mcpu, etc.) to be passed. These can be used for
cross compilation (if supported by the host LLVM installation) and for
selecting specific cpu variants (such as -mcpu=generic which ensures that the
generated code will be portable to all common cpu variants).

** Pure 0.61 2014-09-10

LLVM 3.5 compatibility fixes. Also, for LLVM 3.3 and later, the batch compiler
now uses llc to create native objects directly from the LLVM source. This
should speed up compilation in the batch compiler and also work around bugs
which make llc generate invalid native assembler code on some systems.

** Pure 0.60 2014-03-24


pure/etc/pure.py

# Pure lexer for Pygments (http://pygments.org). This enables syntax
# highlighting for Pure scripts and interactive Pure sessions in Pygments and
# other Pygments-enabled software such as Sphinx.

# Until this becomes part of Pygments, you'll have to install this
# manually. The easiest way to do this is by installing the Pygments addon
# module available here: https://github.com/agraef/pygments-lexer-pure

class PureLexer(RegexLexer):
    """
    For the Pure language.
    """

    name = 'Pure'
    aliases = ['pure']
    filenames = ['*.pure']
    mimetypes = ['text/x-pure']

    keywords = ['infix[lr]?', 'outfix', 'prefix', 'postfix', 'nonfix',
                'namespace', 'private', 'public', 'case', 'const', 'def',
                'else', 'end', 'extern', 'if', 'let', 'of', 'otherwise',
                'then', 'using', 'when', 'with', 'type', 'interface']

    # These aren't really keywords but we want them to stick out anyway.
    keywords2 = ['catch', 'throw', '__break__', '__trace__']

    primitives = ['bigint', 'bool', 'char', 'short', 'int', 'long',
                  'float', 'double', 'expr', 'string', 'pointer', 'void',
                  'int8', 'int16', 'int32', 'int64',
                  'matrix', 'dmatrix', 'cmatrix', 'imatrix',
                  'smatrix', 'nmatrix']

    commands = ['break', 'bt', 'cd', 'clear', 'del', 'dump', 'help', 'ls',
                'mem', 'override', 'pwd', 'quit', 'run', 'save', 'show',
                'stats', 'trace', 'underride']

    tokens = {
        'root': [
            # Whitespace and comments.
            (r'\s+', Text),
            (r'/[*](.|\n)*?[*]/', Comment.Multiline),
            (r'//.*?\n', Comment.Single),
            (r'^#!.*?\n', Comment.Preproc),

            # Inline foreign language code. Avoid highlighting these.
            (r'%<(.|\n)*?%>', Text),

            # Interactive prompts and error messages.
            (r'^> ?$', Generic.Prompt), # lone prompt at end of snippet
            (r'^> ', Generic.Prompt, 'prompt'),
            (r'^[^"\s][^,\n]*, line [0-9]+: ', Generic.Error, 'error'),

            # Various kinds of keywords and built-in punctuation.
            (r'\b(%s)\b' % '|'.join(keywords), Keyword.Reserved),
            (r'\b(%s)\b' % '|'.join(keywords2), Name.Builtin),
            (r'\b(%s)\b' % '|'.join(primitives), Keyword.Type),
            (r'[@;]|::', Punctuation),

            # Numbers.
            (r'\b(\d*\.\d+|(\d+\.\d*|\.\d+|\d+)[eE][+-]?\d+)\b', Number.Float),
            (r'\b(0x[0-9a-fA-F]+L?)\b', Number.Hex),
            (r'\b(0[0-7]+L?)\b', Number.Oct),
            (r'\b(0[bB][01]+L?)\b', Number.Binary),
            (r'\b(\d+L?)\b', Number.Integer),

            # Strings.
            (r'"', String, 'string'),

            # Specials (thunks, closures, pointers, etc.).
            (r'#<[^>\n]+>', Keyword.Type),

            # Pretty much everything else can be a legal (utf-8) identifier or
            # operator symbol in Pure, we just highlight these as text for now.
            (r'\S', Text),
        ],
        'string': [
            (r'"', String, '#pop'),
            (r'\\([\\abfnrtv"\']|&[^;]+;|\([^)]+\))', String.Escape),
            (r'\\(0x[0-9a-fA-F]+|0[0-7]+|0[bB][01]+|\d+)', String.Escape),
            (r'[^\\"\n]+', String), # all other characters
            (r'\\\n', String), # line continuation
            (r'\\', String), # stray backslash
        ],
        'prompt': [
            # Interactive interpreter commands. No highlighting.
            (r'(%s)\b.*?\n' % '|'.join(commands), Text, '#pop'),
            # Interactively typed pragmas.
            (r'#!.*?\n', Comment.Preproc, '#pop'),
            # Anything else is highlighted as usual.
            ('', Text, '#pop'),
        ],
        'error': [
            (r'.*?\n', Text, '#pop'),
        ]
    }
